playback/src/player.rs

```diff
@@ -145,33 +165,44 @@ impl PlayerState {
     fn decoder(&mut self) -> Option<&mut Decoder> {
         use self::PlayerState::*;
         match *self {
-            Stopped => None,
+            Stopped | EndOfTrack { .. } => None,
             Paused { ref mut decoder, .. } |
             Playing { ref mut decoder, .. } => Some(decoder),
             Invalid => panic!("invalid state"),
         }
     }
 
-    fn signal_end_of_track(self) {
+    fn send_end_of_track(self) {
         use self::PlayerState::*;
         match self {
             Paused { end_of_track, .. } |
             Playing { end_of_track, .. } => {
                 let _ = end_of_track.send(());
-            }
+            },
+            _ => ()
+        }
+    }
 
-            Stopped => warn!("signal_end_of_track from stopped state"),
-            Invalid => panic!("invalid state"),
+    fn playing_to_end_of_track(&mut self) {
+        use self::PlayerState::*;
+        match *self {
+            Paused { track_id, .. } |
+            Playing { track_id, .. } => {
+                let old_state = mem::replace(self, EndOfTrack { track_id });
+                old_state.send_end_of_track();
```

Comment:
You should just merge the `send_end_of_track` function into here, so you only have to match once.
You'll probably have to do the `::std::mem::replace(self, Invalid)` "dance" in other to get the `end_of_track` by value. (See `paused_to_playing` for example).

Reply:
Ahh, I was wondering if there was a more elegant way to do that! It didn't occur to me to pattern match on the result of  `replace(self, Invalid)`.

Reply:
Hopefully rust will get NLL soon and we can get rid of all this awkward `Invalid` stuff